content/blog/2022-03-21-conservative-gc/index.md

```diff
@@ -0,0 +1,256 @@
+the authors introduce the concept of an optimized object map, which
+tracks live objects, to mitigate some of these issues. Finally, they compare
+conservative collectors to their exact counterparts and conclude that
+conservative collectors can be as efficient in space and time as exact
+collectors.
+
+# Background
+
+Within language implementations, memory management can be done by
+either _exact_ garbage collectors, which are fully aware of all
+references, or _conservative_ garbage collectors, which must deal with
+_ambiguous references_ that could be either pointers or values.
+
+Because exact collectors know references, they can move objects and
+filter out more references. However, exact collectors impose
+non-trivial engineering and performance challenges, such as having to
+maintain a shadow stack. In many situations, it is also impossible to
+implement an exact collector. So, work on implementing effective and
+efficient conservative collectors is necessary.
+
+The key challenges in implementing conservative collectors are
+as follows:
+
+1. **Excess retention**: Because an ambiguous reference _could_ point to
+an object, that "referent" cannot be collected. This is because the
+ambiguous reference _could_ be a valid pointer to that object. So,
+actually dead objects and their transitively reachable descendants
+will be kept alive, causing extra space to be taken.
+
+2. **Pinning**: Because an ambiguous reference _could_ be a value, the
+collector can't modify it. So, in the case that the ambiguous
+reference is actually a pointer, the referents cannot be moved,
+causing fragmentation.
+
+# Contributions
+
+The main contributions of this paper include:
+
+- A detailed examination on conservative garbage collection, including
+  the first detailed study of impacts of exact and conservative
+  collection in practice.
+
+- Introducing an object map that precisely tracks live objects to filter
+  ambiguous roots efficiently.
+
+- The design, implementation, and evaluation of new conservative
+  collectors that use the object map, against prior conservative
+  collectors and their exact counterparts.
+
+  - **RC Immix<sub>Cons</sub>**, introduced in this paper, is 1%
+  _faster_ than **Gen Immix**, the best-performing exact collector.
+
+<center>
+<img src="fig1.png" alt="Exact versus Conservative Collectors" width="50%" />
+</center>
+
+# Techniques
+
+There are two broad classes of techniques used by conservative collectors to
+make them match exact collectors:
+
+1. **Filtering**: maintaining a data structure to track references to live objects
+   to filter ambiguous references that could not possibly be valid pointers to reduce
+   excess retention.
+
+2. **Ambiguous Roots**: many collectors strike the balance between exact and conservative
+   collection by using available run-time information such that the only conservative references
+   are the roots. Other references are exactly known, and therefore can be moved freely.
```

Comment:
Not sure, but maybe one tiny extra bit of detail here about why the "conservative roots, exact heap" setup such a common hybrid would be helpful context?

Reply:
Added some more text to this effect.